### PR TITLE
Scanner changes

### PR DESCRIPTION
**Pull request by nitin-deamon**
There are 2 commits in this PR.

1. Commit should go to azcopy repo as it export the different APIs.
2. Commit where a sample scanner.go is there which utilize these exported APIs to give scanner functionality. We are utilizing azcopy creation of crawler code which is linked to traverser interface depending on type of source directory.

Need to try with /dev/NULL as destination if it works we can provide dummy destination to /dev/NULL. 
Right now i tried with giving src-> local filesystem, dest-> blob and vice-versa.

### CODE REVIEW

**Comment by shalinijoshi19** [comment]
@nitin-deamon could you also include @johnmic in these PRs so he can build some context etc? Thanks!

**Comment by nitin-deamon** [comment]
> @nitin-deamon could you also include @johnmic in these PRs so he can build some context etc? Thanks!

done @shalinijoshi19

**Comment by zezha-msft** [comment]
Could you please add mohsha-msft nakulkar-msft adreed-msft as reviewers and to this repo? Thanks

**Comment by nitin-deamon** [comment]
> Could you please add mohsha-msft nakulkar-msft adreed-msft as reviewers and to this repo? Thanks

Hi @zezha-msft I added them to repo. Once they accept the invitation , i will added them as reviewer.

File: cmd/zc_enumerator.go [hunk 1]
```diff
@@ -46,7 +46,7 @@ import (
 // represent a local or remote resource object (ex: local file, blob, etc.)
 // we can add more properties if needed, as this is easily extensible
 // ** DO NOT instantiate directly, always use newStoredObject ** (to make sure its fully populated and any preprocessor method runs)
-type storedObject struct {
+type StoredObject struct {
 	name             string
```
Comment (zezha-msft) [inline cmd/zc_enumerator.go:50]: Were these properties missed?

File: scanner/scanner.go [hunk 2]
```diff
@@ -0,0 +1,361 @@
+// Copyright © 2017 Microsoft <nitinsingla@microsoft.com>
```
Comment (zezha-msft) [inline scanner/scanner.go:1]: Minor: please remove the date 2017, for new files we don't need to put a date.

File: scanner/scanner.go [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package scanner
+
+import (
+        "fmt"
```
Comment (zezha-msft) [inline scanner/scanner.go:24]: Please set up `go fmt` to run on file save in your IDE.

File: scanner/scanner.go [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+		//dstRelPath := cca.MakeEscapedRelativePath(false, isDestDir, object)
+                message := fmt.Sprintf("ScannerRun: %v", srcRelPath)
+                glcm.Info(message)
+                return nil
+	}
+	finalizer := func() error {
+		return nil
+	}
+
+	return cmd.NewCopyEnumerator(traverser, filters, processor, finalizer), nil
```
Comment (zezha-msft) [inline scanner/scanner.go:328]: why not use the existing initEnumerator function?

File: cmd/copyEnumeratorInit.go [hunk 5]
```diff
@@ -26,73 +26,73 @@ type BucketToContainerNameResolver interface {
 	ResolveName(bucketName string) (string, error)
 }
 
-func (cca *cookedCopyCmdArgs) initEnumerator(jobPartOrder common.CopyJobPartOrderRequest, ctx context.Context) (*copyEnumerator, error) {
-	var traverser resourceTraverser
+func (cca *CookedCopyCmdArgs) initEnumerator(jobPartOrder common.CopyJobPartOrderRequest, ctx context.Context) (*CopyEnumerator, error) {
```
Comment (zezha-msft) [inline cmd/copyEnumeratorInit.go:29]: This should be useful for you too I think

File: cmd/zc_enumerator.go [hunk 6]
```diff
@@ -587,14 +587,14 @@ type preFilterProvider interface {
 type syncEnumerator struct {
```
Comment (zezha-msft) [inline cmd/zc_enumerator.go:587]: I think these enumerators should be useful too.

**Review by zezha-msft — COMMENTED** [review]
To set the scope for the refactoring, we should clarify the intended usage. There are a couple of choices:
1. Put in raw inputs with rawCopyCmdArgs and invoke the whole execution chain, this is essentially running AzCopy in the same process. Doing so means using AzCopy's schema for saving job plan files and job tracking. This is probably not what you want since the lifecycle manager will call os.exit when the job finishes.
2. Leverage the existing enumerator to traverse, filter, and process stored objects in a customized way that you'd like. Doing so requires all the constructs to be exported. Several of them are missing right now. 
3. Only leverage certain traverser, filter to build your own enumerator or different concept entirely.

File: scanner/scanner.go (hunk 3 above)
Comment (nitin-deamon) [reply scanner/scanner.go:24]: sure Ze, i modified the file using vi, let me start using VSC and add go fmt to run on file save.

File: scanner/scanner.go (hunk 2 above)
Comment (nitin-deamon) [reply scanner/scanner.go:1]: sure got it Ze, Thanks

File: cmd/zc_enumerator.go (hunk 1 above)
Comment (nitin-deamon) [reply cmd/zc_enumerator.go:50]: No Ze, Till now sample code not using any of its properties. May be as i add more functionality i need those properties. Then i will export them. Or i can write function in cmd which work on these properties.

File: cmd/zc_enumerator.go (hunk 6 above)
Comment (nitin-deamon) [reply cmd/zc_enumerator.go:587]: Sync Enumerator may be not , we are trying to scan source only on the basis of Last modified time. As per initial thought we don't want to scan destination, try to keep some meta-data on premise. Which help in finding delta.

File: scanner/scanner.go (hunk 4 above)
Comment (nitin-deamon) [reply scanner/scanner.go:328]: The reason for not using initEnumerator function is we want Processing function and Finalizer Function should work according to our requirement. That diverge lead me to use initResourceEnumerator and create our own initEnumerator function.

**Comment by shalinijoshi19** [comment]
> There are 2 commits in this PR.
> 
> 1. Commit should go to azcopy repo as it export the different APIs.
> 2. Commit where a sample scanner.go is there which utilize these exported APIs to give scanner functionality. We are utilizing azcopy creation of crawler code which is linked to traverser interface depending on type of source directory.
> 
> Need to try with /dev/NULL as destination if it works we can provide dummy destination to /dev/NULL.
> Right now i tried with giving src-> local filesystem, dest-> blob and vice-versa.

@nitin-deamon do we have an updated set of commits at this point already? My assumption is we were going to split this so that #1 could be a new PR in the azcopy repo and #2 would be merged into our msazure instance?

**Review by shalinijoshi19 — CHANGES_REQUESTED** [review]
Hey @nitin-deamon when do you think we'd be able to close on this one? Are there more changes for us to look at since?

**Comment by nitin-deamon** [comment]
> Hey @nitin-deamon when do you think we'd be able to close on this one? Are there more changes for us to look at since?

@shalinijoshi19 Already created the branch in azcopy repo, and cherry pick the changes to it. I need to push changes to our Repo.

**Pull request closed without merging**